### ISSUE TITLE
DeviceCreationError exceptions in cmdlib

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#528]
Originally reported of Google Code with ID 472.

```
replace-disks calls _CreateBlockDevInner() directly (btw, in two points) and bypasses the cleanup mechanism put in place through the DeviceCreationError exception. Apart from the missing cleanup (which could be a different bug), this has an unpleasant consequence: Errors in creating a disk generate a cryptic message. This an example:

 gnt-instance replace-disks -s test-instance
[...]
Thu May 23 17:53:46 2013  - INFO: Adding storage on NODE_NAME for disk/0
Traceback (most recent call last): 
  File "/usr/sbin/gnt-instance", line 21, in <module>
    sys.exit(main.Main())
  File "/usr/lib/python2.6/dist-packages/ganeti/client/gnt_instance.py", line 1611, in Main
    env_override=_ENV_OVERRIDE)
  File "/usr/lib/python2.6/dist-packages/ganeti/cli.py", line 2532, in GenericMain
    result = func(options, args)   
  File "/usr/lib/python2.6/dist-packages/ganeti/client/gnt_instance.py", line 697, in ReplaceDisks
    SubmitOrSend(op, opts)
  File "/usr/lib/python2.6/dist-packages/ganeti/cli.py", line 2277, in SubmitOrSend
    return SubmitOpCode(op, cl=cl, feedback_fn=feedback_fn, opts=opts)
  File "/usr/lib/python2.6/dist-packages/ganeti/cli.py", line 2254, in SubmitOpCode
    reporter=reporter)
  File "/usr/lib/python2.6/dist-packages/ganeti/cli.py", line 2235, in PollJob
    return GenericPollJob(job_id, _LuxiJobPollCb(cl), reporter)
  File "/usr/lib/python2.6/dist-packages/ganeti/cli.py", line 2057, in GenericPollJob
    errors.MaybeRaise(msg)
  File "/usr/lib/python2.6/dist-packages/ganeti/errors.py", line 526, in MaybeRaise
    raise errcls(*args)
TypeError: __init__() takes exactly 3 arguments (1 given)


This bug should be present in 2.7 (I haven't try to reproduce it, but the code is there).

At least the exception should be caught and transformed into an OpExecError.


```
Originally added on 2013-05-23 18:11:30 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
The TypeError is not just "cryptic" is a mistake in the exception code, that basically masks the real exception. This should definitely be fixed. Then we need to find out:

1) What the original error actually was
2) Whether we can avoid "_CreateBlockDevInner"
3) How to cleanup properly if we can't

Thanks,

Guido

```
Originally added on 2013-05-24 07:03:08 +0000 UTC.

Added Labels: Priority-High
Changed State: Accepted
Added to Milestone: Release2.7

**Comment by Ganeti-Issues-Migrator** [comment]
```
The original error was a DeviceCreationError (sorry if it wasn't clear).

The problem with the exception code is that it doesn't know how to handle a DeviceCreationError exception, which wasn't supposed to be sent across the wire in the first place. Perhaps, e could avoid making DeviceCreationError a child of GenericError; this would fix the encoding problem if such an exception is uncaught again.

```
Originally added on 2013-05-24 10:26:33 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Ah ok, sorry.

Indeed we should catch the error for the calls to _CreateBlockDevInner (at least) and raise an OpExecError (that's the minimum). Later (2.8, 2.9) we can try to move from _CreateBlockDevInner to something more sensible in replacedisks, if possible.

wrt DeviceCreationError, it seems to fail because it has too many arguments and thus fails the wire format. True, making it an Exception would work, but then again we have a convention that errors inside ganeti are genericerrors. We could have a special "InternalError" subclass that if raised outside the LU raises a ProgrammerError, perhaps, over the wire?

Thanks,

Guido

```
Originally added on 2013-05-24 10:49:32 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I like the idea of internal errors.

2.8 should be a good target for cleaning up the partially-created disks. But then we should split this bug.
```
Originally added on 2013-05-24 10:57:04 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Move non-critical bugs scheduled for 2.8 or 2.9 to 2.11, as in those versions only critical bug fixes will be integrated.
```
Originally added on 2013-10-30 09:48:16 +0000 UTC.

Added to Milestone: Release2.11
Removed from Milestone: Release2.8

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-01-09 11:34:54 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Removed the milestone as it's long past. As an effort to migrate all our issues to Github, I'll mark this as "Type-Defect" and let someone from the community pick it up if they are interested. We'd be happy to accept patches about this.

See https://groups.google.com/forum/#!topic/ganeti/b2-eukpsXpk for more details.
```
Originally added on 2016-12-13 11:03:45 +0000 UTC.

Added Labels: Type-Defect
Changed State: PatchesWelcome
Removed from Milestone: Release2.11

**Comment by ge-fa** [comment]
Fixed via 1b3fbef66f0cf87fe0e6c3a49de54ddb94b3fdc1.

new state: should be closed
